test/ruby/test_assignment.rb

```diff
@@ -248,6 +248,17 @@ def test_massign_splat
     a,b,*c = *[*[1,2]]; assert_equal([1,2,[]], [a,b,c])
   end
 
+  def test_massign_optimized_literal_bug_21012
+    singleton_class.attr_reader :a
+    @a = []
+    def (@a).[]=(*args)
```

Comment:
The kind of the receiver seems not essential.

```suggestion
    a = []
    def a.[]=(*args)
```